shape mismatch training

Following INSTRUCTIONS.md. Not changing things between build.py and train.py invocation; I get.
```
Traceback (most recent call last):
  File "train.py", line 257, in <module>
    main()
  File "train.py", line 252, in main
    train_model(config, "training/" + args.name, **model_args)
  File "train.py", line 130, in train_model
    run_name=rum_name + " " + datetime.datetime.now().strftime("%Y%m%d-%H%M%S"),
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\model.py", line 658, in train_model
    self.train_samples = self.setup_sample_training_data(LOG_DIR, self.writer_train)
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\model.py", line 583, in setup_sample_training_data
    data = self.datasets.train.fetch_segment(segment, augment=False)
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\dataset.py", line 717, in fetch_segment
    default_inset=self.DEFAULT_INSET,
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\dataset.py", line 337, in apply
    ), "Reference level shape and data shape not match."
AssertionError: Reference level shape and data shape not match.
```

<details>
<summary> Full output</summary>

<p>
```
$ python train.py rat
   INFO Creating new GPU session with memory growth enabled.
2019-04-18 16:04:41.750992: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
WARNING:tensorflow:From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:40: conv2d (from tensorflow.python.layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.conv2d instead.
WARNING From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:40: conv2d (from tensorflow.python.layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.conv2d instead.
WARNING:tensorflow:From C:\Users\robertc\Documents\src\cacophonyenv\lib\site-packages\tensorflow\python\framework\op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING From C:\Users\robertc\Documents\src\cacophonyenv\lib\site-packages\tensorflow\python\framework\op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:58: batch_normalization (from tensorflow.python.layers.normalization) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.batch_normalization instead.
WARNING From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:58: batch_normalization (from tensorflow.python.layers.normalization) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.batch_normalization instead.
   INFO Thermal convolution output shape: (?, 3, 3, 64)
   INFO Motion convolution output shape: (?, 3, 3, 64)
   INFO Output shape (?, ?, 1152)
WARNING:tensorflow:From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:545: LSTMCell.__init__ (from tensorflow.python.ops.rnn_cell_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This class is equivalent as tf.keras.layers.LSTMCell, and will be replaced by that in Tensorflow 2.0.
WARNING From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:545: LSTMCell.__init__ (from tensorflow.python.ops.rnn_cell_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This class is equivalent as tf.keras.layers.LSTMCell, and will be replaced by that in Tensorflow 2.0.
WARNING:tensorflow:From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:558: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `keras.layers.RNN(cell)`, which is equivalent to this API
WARNING From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:558: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `keras.layers.RNN(cell)`, which is equivalent to this API
WARNING:tensorflow:From C:\Users\robertc\Documents\src\cacophonyenv\lib\site-packages\tensorflow\python\ops\rnn_cell_impl.py:1259: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
WARNING From C:\Users\robertc\Documents\src\cacophonyenv\lib\site-packages\tensorflow\python\ops\rnn_cell_impl.py:1259: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
   INFO lstm output shape: ? x (?, 256)
   INFO lstm state shape: (?, 256, 2)
WARNING:tensorflow:From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:579: dense (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dense instead.
WARNING From C:\Users\robertc\Documents\src\classifier-pipeline\model_crnn.py:579: dense (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dense instead.
WARNING:tensorflow:From C:\Users\robertc\Documents\src\cacophonyenv\lib\site-packages\tensorflow\python\ops\losses\losses_impl.py:209: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
WARNING From C:\Users\robertc\Documents\src\cacophonyenv\lib\site-packages\tensorflow\python\ops\losses\losses_impl.py:209: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
   INFO Training segments: 9.3k
   INFO Validation segments: 2.1k
   INFO Test segments: 1.1k
   INFO Labels: ['bird', 'false-positive', 'hedgehog', 'possum', 'rat']
Training on labels ['bird', 'false-positive', 'hedgehog', 'possum', 'rat']

label                train                validation           test                 (segments/tracks/bins/weight)
bird                 606/32/5/1185.0      74/19/4/353.3        74/19/4/80.5
false-positive       1364/178/11/1185.0   1092/152/10/353.4    300/152/10/235.6
hedgehog             5668/220/34/1185.0   436/29/10/353.4      300/29/10/302.7     
possum               181/21/4/1185.0      141/21/4/353.3       141/21/4/138.2
rat                  1472/307/33/1185.0   314/204/27/353.3     300/204/27/237.4

['bird', 'false-positive', 'hedgehog', 'possum', 'rat']
['bird', 'false-positive', 'hedgehog', 'possum', 'rat']
['bird', 'false-positive', 'hedgehog', 'possum', 'rat']
Training started
---------------------
Hyper parameters
---------------------
augmentation=True
thermal_threshold=10
scale_frequency=0.5
keep_prob=0.2
batch_size=16
learning_rate=0.0001
learning_rate_decay=1.0
l2_reg=0
label_smoothing=0.1
batch_norm=True
lstm_units=256
enable_flow=True

Found 9.3K training examples

Initialising summary writers at D:\Cacophony\logs\training/rat 20190418-160444.
Traceback (most recent call last):
  File "train.py", line 257, in <module>
    main()
  File "train.py", line 252, in main
    train_model(config, "training/" + args.name, **model_args)
  File "train.py", line 130, in train_model
    run_name=rum_name + " " + datetime.datetime.now().strftime("%Y%m%d-%H%M%S"),
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\model.py", line 658, in train_model
    self.train_samples = self.setup_sample_training_data(LOG_DIR, self.writer_train)
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\model.py", line 583, in setup_sample_training_data
    data = self.datasets.train.fetch_segment(segment, augment=False)
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\dataset.py", line 717, in fetch_segment
    default_inset=self.DEFAULT_INSET,
  File "C:\Users\robertc\Documents\src\classifier-pipeline\ml_tools\dataset.py", line 337, in apply
    ), "Reference level shape and data shape not match."
AssertionError: Reference level shape and data shape not match.
(cacophonyenv) 
```

</p>
</details>

--- a/classifier_TEMPLATE.yaml
+++ b/classifier_TEMPLATE.yaml
@@ -156,7 +156,7 @@ train:
     hyper_params:
         # training
         batch_size: 16
-        learning_rate: 1e-4
+        learning_rate: 0.0004
         learning_rate_decay: 1.0
         l2_reg: 0
         label_smoothing: 0.1

--- a/ml_tools/model.py
+++ b/ml_tools/model.py
@@ -16,9 +16,6 @@ from sklearn import metrics
 from ml_tools import tools
 from ml_tools import visualise
 
-# folder to save model while it's training.  Make sure this isn't on a dropbox folder and it will cause a crash.
-CHECKPOINT_FOLDER = "c:\cac\checkpoints"
-
 
 class Model:
     """ Defines a deep learning model """
@@ -27,7 +24,7 @@ class Model:
     MODEL_DESCRIPTION = ""
     VERSION = "0.3.0"
 
-    def __init__(self, session=None):
+    def __init__(self, train_config=None, session=None):
 
         self.name = "model"
         self.session = session or tools.get_session()
@@ -96,7 +93,12 @@ class Model:
         self.enable_async_loading = True
 
         # folder to write tensorboard logs to
-        self.log_dir = "./logs"
+        if train_config:
+            self.log_dir = os.path.join(train_config.train_dir, "logs")
+            self.checkpoint_folder = os.path.join(train_config.train_dir, "checkpoints")
+        else:
+            self.log_dir = "./logs"
+            self.checkpoint_folder = "./checkpoints"
         self.log_id = ""
 
         # number of frames per segment during training
@@ -721,12 +723,12 @@ class Model:
                 )
 
                 # create a save point
-                self.save(os.path.join(CHECKPOINT_FOLDER, "training-most-recent.sav"))
+                self.save(os.path.join(self.checkpoint_folder, "training-most-recent.sav"))
 
                 # save the best model if validation score was good
                 if val_loss < best_val_loss:
                     print("Saving best validation model.")
-                    self.save(os.path.join(CHECKPOINT_FOLDER, "training-best-val.sav"))
+                    self.save(os.path.join(self.checkpoint_folder, "training-best-val.sav"))
                     best_val_loss = val_loss
 
                 # save at epochs
@@ -748,7 +750,7 @@ class Model:
                     self.eval_score = acc
                     self.save(
                         os.path.join(
-                            CHECKPOINT_FOLDER,
+                            self.checkpoint_folder,
                             "training-epoch-{:02d}.sav".format(int(epoch)),
                         )
                     )
@@ -758,7 +760,7 @@ class Model:
                         print("Save best epoch tested model.")
                         # saving a copy in the log dir allows tensorboard to access some additional information such
                         # as the current training data varaibles.
-                        self.save(os.path.join(CHECKPOINT_FOLDER, "training-best.sav"))
+                        self.save(os.path.join(self.checkpoint_folder, "training-best.sav"))
                         try:
                             self.save(
                                 os.path.join(
@@ -789,7 +791,7 @@ class Model:
         # restore previous best
         if self.use_best_weights:
             print("Using model from step", best_step)
-            self.load(os.path.join(CHECKPOINT_FOLDER, "training-best.sav"))
+            self.load(os.path.join(self.checkpoint_folder, "training-best.sav"))
 
         self.eval_score = self.eval_model()
 

--- a/ml_tools/model.py
+++ b/ml_tools/model.py
@@ -723,12 +723,16 @@ class Model:
                 )
 
                 # create a save point
-                self.save(os.path.join(self.checkpoint_folder, "training-most-recent.sav"))
+                self.save(
+                    os.path.join(self.checkpoint_folder, "training-most-recent.sav")
+                )
 
                 # save the best model if validation score was good
                 if val_loss < best_val_loss:
                     print("Saving best validation model.")
-                    self.save(os.path.join(self.checkpoint_folder, "training-best-val.sav"))
+                    self.save(
+                        os.path.join(self.checkpoint_folder, "training-best-val.sav")
+                    )
                     best_val_loss = val_loss
 
                 # save at epochs
@@ -760,7 +764,9 @@ class Model:
                         print("Save best epoch tested model.")
                         # saving a copy in the log dir allows tensorboard to access some additional information such
                         # as the current training data varaibles.
-                        self.save(os.path.join(self.checkpoint_folder, "training-best.sav"))
+                        self.save(
+                            os.path.join(self.checkpoint_folder, "training-best.sav")
+                        )
                         try:
                             self.save(
                                 os.path.join(

--- a/model_crnn.py
+++ b/model_crnn.py
@@ -271,7 +271,7 @@ class ModelCRNN_HQ(ConvModel):
         "scale_frequency": 0.5,
     }
 
-    def __init__(self, labels, **kwargs):
+    def __init__(self, labels, train_config, **kwargs):
         """
         Initialise the model
         :param labels: number of labels for model to predict
@@ -470,7 +470,7 @@ class ModelCRNN_LQ(ConvModel):
         "scale_frequency": 0.5,
     }
 
-    def __init__(self, labels, **kwargs):
+    def __init__(self, labels, train_config, **kwargs):
         """
         Initialise the model
         :param labels: number of labels for model to predict

--- a/track/trackextractor.py
+++ b/track/trackextractor.py
@@ -687,7 +687,7 @@ class TrackExtractor:
     def start_and_end_in_secs(self, track):
         return (
             self.frame_time_in_secs(track, 0),
-            self.frame_time_in_secs(track, len(track) - 1),
+            self.frame_time_in_secs(track, len(track)),
         )
 
     def frame_time_in_secs(self, track, frame_index=0):

--- a/train/train.py
+++ b/train/train.py
@@ -21,11 +21,8 @@ def train_model(run_name, conf, hyper_params):
         dsets = pickle.load(f)
     labels = dsets[0].labels
 
-    model = ModelCRNN_LQ(labels=len(labels), **hyper_params)
-
+    model = ModelCRNN_LQ(labels=len(labels), train_config=conf.train, **hyper_params)
     model.import_dataset(datasets_filename)
-    model.log_dir = os.path.join(conf.train.train_dir, "logs")
-
     # display the data set summary
     print("Training on labels", labels)
     print()

